picky/src/pkcs12/mod.rs

```diff
@@ -0,0 +1,606 @@
+//! This module provides types for parsing and building PKCS#12 files (PFX).
+//!
+//! # PFX file structure
+//! PFX files are constructed from `safe contents` on the highest level, which are containers for
+//! various types of `safe bags`. They could be encrypted or not and contain one or more `safe bags`.
+//!
+//! `Safe bags` are polymorphic containers for various types of data, such as private keys, certificates,
+//! custom data (`secrets`), CRLs, etc. They could also be nested, which means that one `safe bag` could
+//! contain multiple `safe bags` inside, making it possible to represent complex tree-like structures.
+//!
+//! Each `safe bag` could also contain `attributes`, which are key-value pairs of various types of data.
+//! High-level API defines only `friendly name` and `local key id` attributes, but it is possible to
+//! parse/build any custom attributes as well.
+//!
+//! There are also special `safe bags` called `shrouded key bags`, which are basically encrypted
+//! PKCS8 private keys. Encryption on `safe bag` level is performed to allow PFX files that have
+//! unencrypted certificates and encrypted private keys, to be preprocessed by the software without
+//! knowing PFX password (e.g. check if PFX contains private key for the certificate or not). Usually
+//! relations between contained certificates and private keys are marked with `local key id` attribute
+//! which is set to the same value for both certificate and private key.
+//!
+//! # Parsing
+//! Almost any kind of PKCS#12 object specified by [RFC](https://datatracker.ietf.org/doc/html/rfc7292)
+//! could be parsed or constructed. However, only the most used PFX types are wrapped in high level
+//! convenient API such as private keys, certificates, nested pfx nodes, etc. If parsed
+//! data was in fact some rarely used PFX type, it is still possible to access it via low level API
+//! via `.inner()` method usually returns ASN.1 DER encoded data (`picky-asn1-x509` crate types).
+//!
+//! Parsing process could be controlled even more with [`Pkcs12ParsingParams`], which allows to skip
+//! some parsing errors (e.g. Skip failed mapping of private key ASN.1 structure to high-level picky
+//! wrapper)
+//!
+//! Parsing of PFX file is simple as calling [`Pfx::from_der_bytes`] method with required params
```

Comment:
(If you address the other comments, you’ll need to update this as well)

Reply:
Done